### PR TITLE
add dynamic shapes support for subclasses that override size/stride

### PR DESCRIPTION
**Pull request by bdhirsh**
This is mostly a minor fix on top of @soulitzer's PR https://github.com/pytorch/pytorch/pull/107839.

(1) `strides` wasn't going through the new `set_tensor_attr_with_capsule` flow
(2) The dynamic shapes overload for `_make_wrapper_subclass` currently errors when you try to use custom sizes - I removed the error
(3) added a test

I need this later because I'm adding a `__torch_dispatch__` `FunctionalTensor` wrapper subclass, that needs to support dynamic shapes, and also plumb metadata calls to its inner tensor later.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #107917
* #106404
* #107926
* __->__ #107916
* #107915

### CODE REVIEW

File: torch/csrc/autograd/python_variable.cpp [hunk 1]
```diff
@@ -720,9 +720,8 @@ static PyObject* THPVariable_make_wrapper_subclass(
 
     const auto sizes_strides_policy = r.stringViewOptional(10);
     if (sizes_strides_policy.has_value()) {
-      TORCH_CHECK(
-          false,
-          "Setting sizes_strides_policy isn't supported for this overload")
+      tensor.unsafeGetTensorImpl()->set_python_custom_sizes_strides(
+          parseSizesStridesPolicyArgument(*sizes_strides_policy));
```
Comment (bdhirsh) [inline torch/csrc/autograd/python_variable.cpp:724]: I actually don't think I need this, double checking now

File: torch/csrc/PyInterpreter.cpp [hunk 2]
```diff
@@ -565,37 +576,21 @@ c10::IntArrayRef ConcretePyInterpreterVTable::strides(
...
+  int64_t len = py::len(out);
+  int64_t* ptr = new int64_t[len];
+  auto capsule =
+      py::capsule(ptr, [](void* p) { delete[] reinterpret_cast<int64_t*>(p); });
+  int64_t idx = 0;
+  for (auto it = out.begin(); it != out.end(); ++it, ++idx) {
+    ptr[idx] = py::cast<int64_t>(*it);
+  }
+  set_tensor_attr_with_capsule(
+      const_cast<c10::TensorImpl*>(self), capsule, "_sizes_capsule");
```
Comment (soulitzer) [inline torch/csrc/PyInterpreter.cpp:592]:
```suggestion
      const_cast<c10::TensorImpl*>(self), capsule, "_strides_capsule");
```

Comment (bdhirsh) [reply torch/csrc/PyInterpreter.cpp:592]: whoops- thanks!

File: test/test_python_dispatch.py [hunk 3]
```diff
@@ -2028,6 +2028,53 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
...
+                if func == torch.ops.aten.sym_stride.default:
+                    return args[0].inner.shape
+                return NotImplemented
+
+        x = torch.ones(2, 2)
+        def trace_fn(x):
+            x_wrapper = CustomSizeDynamicShapesTensor(x)
+            return x_wrapper.size(), x_wrapper.stride()
+        fx_g = make_fx(trace_fn, tracing_mode="symbolic")(x)
+        # This is broken
```
Comment (ezyang) [inline test/test_python_dispatch.py:2071]: Is it? Seems like it does what the torch dispatch said it should do?

**Review by ezyang — APPROVED** [review]

File: test/test_python_dispatch.py (hunk 3 above)
Comment (bdhirsh) [reply test/test_python_dispatch.py:2071]: **was** broken 🙂. Forgot to remove (during local testing)

**Review by soulitzer — APPROVED** [review]
Thanks!

File: test/test_python_dispatch.py [hunk 4]
```diff
@@ -2028,6 +2028,53 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
...
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args, kwargs):
+                if func == torch.ops.aten.sym_size.default:
+                    return args[0].inner.shape
+                if func == torch.ops.aten.sym_stride.default:
+                    return args[0].inner.shape
+                return NotImplemented
+
+        x = torch.ones(2, 2)
+        def trace_fn(x):
```
Comment (soulitzer) [inline test/test_python_dispatch.py:2067]: lint

**Pull request closed without merging**